### PR TITLE
Add theme support to mimetypeIcon through imagePath integration

### PR DESCRIPTION
**Pull request by Egari**
Added theme support to mimetype icons by using the pre-existing imagePath function

### CODE REVIEW

**Comment by Egari** [comment]
This contribution may be used under the MIT license.

**Comment by karlitschek** [comment]
@jancborchardt @DeepDiver1975 What do you think?

**Comment by jancborchardt** [comment]
So this fixes https://github.com/owncloud/core/issues/15320?

@krempelweb @FunkyM can you test this?

**Comment by jancborchardt** [comment]
@Egari also, is this similar to a fix for https://github.com/owncloud/core/issues/6773 maybe? :)

**Comment by Egari** [comment]
@jancborchardt yes it will indeed fix #15320. 
Regarding #6773, in those cases the app tries to load images using the pure-javascript OC.imagePath and OC.filePath functions, which will not use any images and/or files that might be available in your theme. Since the pure-javascript OC.imagePath function does not actually use serverside code, it's more difficult to check for existing files. A possible change for this is to let apps do an ajax call to a PHP server-side url generator (or OC_Helper in this case). 
So in short, unfortunately no.

**Comment by MorrisJobke** [comment]
@Egari Can I ask you to use tabs for intendation? See https://doc.owncloud.org/server/8.1/developer_manual/general/codingguidelines.html#coding

This also would make the diff more pretty.

Thanks

**Comment by Egari** [comment]
@MorrisJobke Of course, my apologies. Updated.

**Comment by nickvergessen** [comment]
@DeepDiver1975 8.1?

**Comment by DeepDiver1975** [comment]
@Egari can I ask you to squash the commits? THX

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 8.1?

hmm .... better not -> 8.2

**Comment by jancborchardt** [comment]
I’d also say 8.2 since it has the potential to break stuff.

**Comment by Egari** [comment]
@DeepDiver1975 squashed into single commit, sorry for the delay.

File: lib/private/helper.php [hunk 1]
```diff
@@ -268,34 +268,42 @@ public static function mimetypeIcon($mimetype) {
...
 		}
 
 		// Icon exists?
-		if (file_exists(OC::$SERVERROOT . '/core/img/filetypes/' . $icon . '.png')) {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/' . $icon . '.png';
-			return OC::$WEBROOT . '/core/img/filetypes/' . $icon . '.png';
+		if (file_exists(OC::$SERVERROOT . '/core/img/filetypes/' . $icon . '.png') ||
+			file_exists(OC::$SERVERROOT . '/core/img/filetypes/' . $icon . '.svg') ||
+			file_exists(OC::$SERVERROOT . '/themes/'.$theme.'/core/img/filetypes/'.$icon.'.png') ||
+			file_exists(OC::$SERVERROOT . '/themes/'.$theme.'/core/img/filetypes/'.$icon.'.svg')) {
```
Comment (Xenopathic) [inline lib/private/helper.php]: This is effectively duplicated code from UrlGenerator (which is where `self::imagePath()` goes). Instead of doing all the `file_exists()` stuff here, just call `self::imagePath()`, then if there is a result it must exist (as per `UrlGenerator::imagePath()`). Note that `imagePath()` will throw a RuntimeException if the image can't be found, which provides a nice clean way of implementing this:

```
try {
// Icon exists?
self::$mimetypeIcons[$mimetype] = self::imagePath(...);
return self::$mimetypeIcons[$mimetype];
} catch (\RuntimeException $e) {
}
try {
// Try only the first part of the filetype
self::$mimetypeIcons[$mimetype] = self::imagePath(...);
return self::$mimetypeIcons[$mimetype];
} catch (\RuntimeException $e) {
}
```

Comment (Egari) [inline lib/private/helper.php]: If Exception-catching (something I'm personally not a fan of) is preferred over code duplication, I will update the pull request when I get the chance.

Comment (Xenopathic) [inline lib/private/helper.php]: Code duplication in itself isn't necessarily a bad thing, but when code that has the identical function is split over two files, then it is a recipe for disaster and inconsistency. The exception plumbing is already there in the `imagePath()` function, and it'd be silly to just ignore it.

**Comment by Xenopathic** [comment]
:+1:

File: lib/private/helper.php [hunk 2]
```diff
@@ -264,39 +264,44 @@ public static function mimetypeIcon($mimetype) {
...
-		if (file_exists(OC::$SERVERROOT . '/core/img/filetypes/' . $icon . '.png')) {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/' . $icon . '.png';
-			return OC::$WEBROOT . '/core/img/filetypes/' . $icon . '.png';
-		}
+		try {
+			self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/' . $icon . '.svg');
+			return self::$mimetypeIcons[$mimetype];
+		} catch (\RuntimeException $e) {
+            // Specified image not found
+        }
```
Comment (nickvergessen) [inline lib/private/helper.php]: intendation is off

File: lib/private/helper.php [hunk 3]
```diff
@@ -264,39 +264,44 @@ public static function mimetypeIcon($mimetype) {
...
-		} else {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/file.png';
-			return OC::$WEBROOT . '/core/img/filetypes/file.png';
-		}
+		try {
+			self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/' . $mimePart . '.svg');
+			return self::$mimetypeIcons[$mimetype];
+		} catch (\RuntimeException $e) {
+            // Image for the first part of the mimetype not found
+        }
```
Comment (nickvergessen) [inline lib/private/helper.php]: intendation is off

File: lib/private/helper.php [hunk 4]
```diff
@@ -264,39 +264,44 @@ public static function mimetypeIcon($mimetype) {
 		if (isset(self::$mimetypeIcons[$mimetype])) {
 			return self::$mimetypeIcons[$mimetype];
 		}
-		// Replace slash and backslash with a minus
+
+        // Replace slash and backslash with a minus
```
Comment (nickvergessen) [inline lib/private/helper.php]: intendation is off

Comment (nickvergessen) [inline lib/private/helper.php]: okay actually you use spaces instead of tabs

**Comment by Egari** [comment]
@nickvergessen Darn, I'll .vimrc the thing... I have it set to spaces by default, and I mustve forgotten to noexpandtab... Fixed code.

**Comment by nickvergessen** [comment]
Thanks, sorry for the troubles

**Comment by jancborchardt** [comment]
@vgiannoul Please review this as it probably fixes your https://github.com/owncloud/core/issues/16547

**Comment by Xenopathic** [comment]
My :+1: still stands

**Comment by vgiannoul** [comment]
maybe I'm wrong or this is my fault but I get an error when applying this patch

Fatal error: Call to undefined method OC_Helper::userAvatarSet() in owncloud/lib/private/templatelayout.php on line 80 

and I have to comment out this line in order to fix it. I have enabled userAvatars in my configuration

**Comment by Egari** [comment]
That function is defined not 20 lines below the end of the changed function;
https://github.com/Egari/core/blob/addThemeSupportToMimeIcon/lib/private/helper.php#L327

I can only imagine something went wrong applying the patch?

**Comment by vgiannoul** [comment]
Sorry for late response, 
i continue to get this error message
Fatal error: Class 'OC\Avatar' not found in /Users/vgiannoul/Sites/owncloud/lib/private/helper.php on line 328
after applying the patch on a clean install (version 8.0.3)

**Comment by DeepDiver1975** [comment]
This patch can only be applied against master.

**Comment by PVince81** [comment]
@rullzer what do you think ? (in regard to your other changes regarding mimetypes)

**Comment by PVince81** [comment]
@owncloud-bot ok to test

File: lib/private/helper.php [hunk 5]
```diff
@@ -264,39 +264,44 @@ public static function mimetypeIcon($mimetype) {
...
+
 		// Replace slash and backslash with a minus
 		$icon = str_replace('/', '-', $mimetype);
 		$icon = str_replace('\\', '-', $icon);
 
 		// Is it a dir?
 		if ($mimetype === 'dir') {
-			self::$mimetypeIcons[$mimetype] = OC::$WEBROOT . '/core/img/filetypes/folder.png';
-			return OC::$WEBROOT . '/core/img/filetypes/folder.png';
+			self::$mimetypeIcons[$mimetype] = self::imagePath('core', 'filetypes/folder.svg');
```
Comment (rullzer) [inline lib/private/helper.php]: I can't check but moving to SVG here by default might break some browsers..

Comment (rullzer) [inline lib/private/helper.php]: O and don't use self::imagePath!!! That function is deprecated.
Use \OC::$server->getURLGenerator()->imagePath($app, $image)

Comment (rullzer) [inline lib/private/helper.php]: This of course hold in the other cases as well

Comment (Egari) [inline lib/private/helper.php]: You're quite right, updated to remove imagePath and revert the change to use svg by default.
Thanks for the feedback! :+1:

Comment (rullzer) [inline lib/private/helper.php]: I would love to drop the PNGs... but unfortunatly we can't.... (yet)

**Comment by rullzer** [comment]
Looking good and working properly for themes.
I still think this function needs to be migrated (#15543). But this theming stuff should be fixed!

:+1:

**Comment by MorrisJobke** [comment]
Tested and works :+1:
